src/rulesets/LinearAlgebra/dense.jl

```diff
@@ -7,14 +7,16 @@ function frule((_, Δx, Δy), ::typeof(dot), x, y)
 end
 
 function rrule(::typeof(dot), x::AbstractArray, y::AbstractArray)
+    project_x = ProjectTo(x)
+    project_y = ProjectTo(y)
     function dot_pullback(Ω̄)
         ΔΩ = unthunk(Ω̄)
         xthunk = InplaceableThunk(
-            @thunk(reshape(y .* ΔΩ', axes(x))),
+            @thunk(project_x(reshape(y .* ΔΩ', axes(x)))),
             dx -> dx .+= reshape(y, axes(x)) .* ΔΩ',
```

Comment:
The other design issue I meant to raise is this: if x is real and y is complex, then this in-place rule will give an InexactError. I believe you probably want to insert something like `eltype(project_x)` into that broadcast. Which seems much tricker to do automatically.

Non-broadcasting in-place rules like `mul!` would need something similar, but not quite so simple.